compiler/vm/vmchecks.nim

```diff
@@ -97,22 +97,29 @@ func testLocationType(locType: PVmType, t: PVmType): AccessViolationReason =
 
 
 func checkValid(p: VmMemPointer, typ: PVmType, cell: VmCell): AccessViolationReason =
-  assert p.rawPointer >= cell.p, "memory not part of `cell`"
-
-  let off = cast[int](p) - cast[int](cell.p)
-  if off > 0:
+  ## Tests whether `p` points to a valid memory location inside `cell` that is
+  ## accesible as type `typ`, and returns either a success or the reason for
+  ## why an access through `p` would be illegal.
+  if p.rawPointer > cell.p:
     # `p` is an interior pointer
-    if typ != nil:
+    let off = cast[int](p) - cast[int](cell.p)
+    if off >= cell.sizeInBytes:
+      # `p` doesn't point inside the cell
+      result = avrOutOfBounds
+    elif typ != nil:
       # the cell stores either a single- or a contiguous sequence of locations
       let stride = cell.typ.alignedSize.int
       result = searchInArray(toOpenArray(cell.p, 0, int(cell.sizeInBytes-1)), cell.typ, typ, cell.count, stride, off)
     else:
       # untyped memory; not yet supported
       unreachable()
-  else:
+  elif p.rawPointer == cell.p:
```

Comment:
I'd personally lead with this branch as it's the most straightforward case and for the reader eliminates that, then they're primed to think about edge cases (greater and less than).

(very soft suggestion; for example, if you need this change in sooner and don't want to bother with the CI time, I'd ignore it)

Reply:
I reordered the branches -- it definitely reads better now.